pkg/apis/pipeline/v1beta1/pipeline_types.go

```diff
@@ -165,6 +170,80 @@ type PipelineTask struct {
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
 }
 
+// validateRefOrSpec validates at least one of taskRef or taskSpec is specified
+func (pt PipelineTask) validateRefOrSpec() (errs *apis.FieldError) {
+	// can't have both taskRef and taskSpec at the same time
+	if pt.TaskRef != nil && pt.TaskSpec != nil {
+		errs = errs.Also(apis.ErrMultipleOneOf("taskRef", "taskSpec"))
+	}
+	// Check that one of TaskRef and TaskSpec is present
+	if pt.TaskRef == nil && pt.TaskSpec == nil {
+		errs = errs.Also(apis.ErrMissingOneOf("taskRef", "taskSpec"))
+	}
+	return errs
+}
+
+// validateCustomTask validates custom task specifications - checking kind and fail if not yet supported features specified
+func (pt PipelineTask) validateCustomTask() (errs *apis.FieldError) {
+	if pt.TaskRef.Kind == "" {
+		errs = errs.Also(apis.ErrInvalidValue("custom task ref must specify kind", "taskRef.kind"))
+	}
+	// Conditions are deprecated so the effort to support them with custom tasks is not justified.
+	// When expressions should be used instead.
+	if len(pt.Conditions) > 0 {
+		errs = errs.Also(apis.ErrInvalidValue("custom tasks do not support conditions - use when expressions instead", "conditions"))
+	}
+	// TODO(#3133): Support these features if possible.
+	if pt.Retries > 0 {
+		errs = errs.Also(apis.ErrInvalidValue("custom tasks do not support retries", "retries"))
+	}
+	if pt.Resources != nil {
+		errs = errs.Also(apis.ErrInvalidValue("custom tasks do not support PipelineResources", "resources"))
+	}
+	if pt.Timeout != nil {
+		errs = errs.Also(apis.ErrInvalidValue("custom tasks do not support timeout", "timeout"))
+	}
+	return errs
+}
+
+// validateBundle validates bundle specifications - checking name and bundle
+func (pt PipelineTask) validateBundle() (errs *apis.FieldError) {
+	// bundle requires a TaskRef to be specified
+	if (pt.TaskRef != nil && pt.TaskRef.Bundle != "") && pt.TaskRef.Name == "" {
+		errs = errs.Also(apis.ErrMissingField("taskRef.name"))
+	}
+	// If a bundle url is specified, ensure it is parsable
+	if pt.TaskRef != nil && pt.TaskRef.Bundle != "" {
+		if _, err := name.ParseReference(pt.TaskRef.Bundle); err != nil {
+			errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("invalid bundle reference (%s)", err.Error()), "taskRef.bundle"))
+		}
+	}
+	return errs
+}
+
+// validateRegularTask validates regular pipeline task (dag/final) for taskRef and taskSpec
+func (pt PipelineTask) validateRegularTask(ctx context.Context) (errs *apis.FieldError) {
```

Comment:
One small nit I have is that the meaning of "Regular" is a bit ambiguous and could change meaning over time.

I wonder if maybe just `validateTask` would be ok for now instead?

Reply:
Done, thanks @sbwsg